### PR TITLE
Activity local dispatch

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Support activity local dispatch: an optimization for skipping matching and task polling for running activity tasks.
- When responding workflow task complete, sdk can set `request_start` to true for an activity command. If server accepts (currently server always accepts if the feature is enabled for that namespace) the request, then the activity will be scheduled AND started, and be returned to the sdk in respondWorkflowTaskResponse. 
- SDK should guarantee that if `request_start` is true, it's has the capacity and is able to run the activity (even if it's cross-namespace or on a different task list)
- Server will guarantee only those activities requested to be started will be included in respondWorkflowTaskResponse. 
- API change: https://github.com/temporalio/api/pull/159

**Why?**
- Activity local dispatch

**How did you test it?**
- Added unit test

**Potential risks**
- No, feature disabled by default.

**Is hotfix candidate?**
- No.

### CODE REVIEW

File: service/history/workflowTaskHandler.go [hunk 1]
```diff
@@ -229,19 +242,87 @@ func (handler *workflowTaskHandlerImpl) handleCommandScheduleActivity(
...
-		return err
+		return nil, err
 	}
-	return nil
+
+	if !localDispatchActivity {
+		return nil, nil
+	}
+
+	if _, err := handler.mutableState.AddActivityTaskStartedEvent(
```
Comment (yiminc) [inline service/history/workflowTaskHandler.go:274]: does server control how many activities are allowed for local dispatch? or all of them as SDK asked?

File: service/history/workflowTaskHandler.go [hunk 2]
```diff
@@ -229,19 +242,87 @@ func (handler *workflowTaskHandlerImpl) handleCommandScheduleActivity(
...
+			ScheduleId:      event.EventId,
+			ScheduleAttempt: ai.Attempt,
+			ActivityId:      attr.ActivityId,
+			ActivityType:    attr.ActivityType.GetName(),
+		}
+		serializedToken, err := handler.tokenSerializer.Serialize(taskToken)
+		if err != nil {
+			return err
+		}
+		response := &workflowservice.PollActivityTaskQueueResponse{
```
Comment (yiminc) [inline service/history/workflowTaskHandler.go:299]: rename response to activityTask

File: service/history/workflowTaskHandler.go [hunk 3]
```diff
@@ -229,19 +242,87 @@ func (handler *workflowTaskHandlerImpl) handleCommandScheduleActivity(
...
+			StartToCloseTimeout:         attr.StartToCloseTimeout,
+			HeartbeatTimeout:            attr.HeartbeatTimeout,
+			TaskToken:                   serializedToken,
+			Attempt:                     ai.Attempt,
+			HeartbeatDetails:            ai.LastHeartbeatDetails,
+			WorkflowType:                handler.mutableState.GetWorkflowType(),
+			WorkflowNamespace:           namespace,
+
+			// RetryPolicy: TODO: do we need this field?
+			// not specified by matching or frontend as well
```
Comment (yiminc) [inline service/history/workflowTaskHandler.go:321]: it seems this is not used by sdk (checked sdk-go). I think we should not have this field.

File: service/history/workflowTaskHandler.go (hunk 1 above)
Comment (yycptt) [reply service/history/workflowTaskHandler.go:275]: Synced offline. Currently there's no obvious reason for server to reject the request (even if the activity is scheduled cross namespace or on a different task queue). The workflow mutable state update transaction and history have the same size.

If worker set `request_start` to true, then it should guarantee it is able to run that activity.

File: service/history/workflowTaskHandler.go (hunk 3 above)
Comment (yycptt) [reply service/history/workflowTaskHandler.go]: Removed the TODO. I will double check with SDK side before removing it from proto def.

**Review by yiminc — APPROVED** [review]

**Pull request merged**